krm-functions/lib/nad/v1/nad.go

```diff
@@ -0,0 +1,364 @@
+		return nadConfigStruct, fmt.Errorf("invalid NAD Config, %s", err)
+	}
+	nadConfigStruct.CniVersion = CniVersion
+	if r.CniSpecType == VlanAllocOnly {
+		return nadConfigStruct, nil
+	} else if r.CniSpecType == IpVlanType {
+		if nadConfigStruct.Plugins == nil || len(nadConfigStruct.Plugins) == 0 {
+			nadConfigStruct.Plugins = []PluginCniType{
+				{
+					Capabilities: Capabilities{
+						Ips: true,
+					},
+					Mode: ModeL2,
+					Ipam: Ipam{
+						Type: NadType,
+						Addresses: []Addresses{
+							{},
+						},
+					},
+				},
+			}
+		}
+		return nadConfigStruct, nil
+	} else if r.CniSpecType == MacVlanType {
+		if nadConfigStruct.Plugins == nil || len(nadConfigStruct.Plugins) == 0 {
+			nadConfigStruct.Plugins = []PluginCniType{
+				{
+					Capabilities: Capabilities{Ips: true},
+					Mode:         ModeBridge,
+					Ipam: Ipam{
+						Type: NadType,
+						Addresses: []Addresses{
+							{},
+						},
+					},
+				},
+				{
+					Capabilities: Capabilities{Mac: true},
+					Type:         TuningType,
+				},
+			}
+		}
+		return nadConfigStruct, nil
+	} else if r.CniSpecType == SriovType {
+		if nadConfigStruct.Plugins == nil || len(nadConfigStruct.Plugins) == 0 {
+			nadConfigStruct.Plugins = []PluginCniType{
+				{
+					Capabilities: Capabilities{Ips: true},
+					Mode:         ModeBridge,
+					Ipam: Ipam{
+						Type: NadType,
+						Addresses: []Addresses{
+							{},
+						},
+					},
+				},
+			}
+		}
+		return nadConfigStruct, nil
+	} else if r.CniSpecType == BothIpamVlan {
+		if nadConfigStruct.Plugins == nil || len(nadConfigStruct.Plugins) == 0 {
+			nadConfigStruct.Plugins = []PluginCniType{
+				{
+					Capabilities: Capabilities{
+						Ips: true,
+					},
+					Mode: ModeBridge,
+					Ipam: Ipam{
+						Type: NadType,
+						Addresses: []Addresses{
+							{},
+						},
+					},
+				},
+			}
+		}
+		return nadConfigStruct, nil
+	}
+	return nadConfigStruct, nil
+}
+
+// SetConfigSpec sets the spec attributes in the kubeobject according the go struct
+func (r *NadStruct) SetConfigSpec(spec *nadv1.NetworkAttachmentDefinitionSpec) error {
+	return r.K.SetNestedString(spec.Config, ConfigType...)
+}
+
+func (r *NadStruct) SetNadConfig(config NadConfig) error {
+	b, err := json.Marshal(config)
+	if err != nil {
+		return err
+	}
+	return r.K.SetNestedString(string(b), ConfigType...)
+}
+
+func (r *NadStruct) SetCNIType(cniType string) error {
+	if cniType != "" {
+		if cniType == "ipvlan" {
+			r.CniSpecType = IpVlanType
+		} else if cniType == "macvlan" {
+			r.CniSpecType = MacVlanType
+		} else if cniType == "sriov" {
+			r.CniSpecType = SriovType
+		}
+		nadConfigStruct, err := r.GetNewNadConfig()
+		if err != nil {
+			return err
+		}
+		for i, plugin := range nadConfigStruct.Plugins {
+			if plugin.Type == TuningType {
+				continue
+			} else {
+				nadConfigStruct.Plugins[i].Type = cniType
+			}
+		}
+		return r.SetNadConfig(nadConfigStruct)
+	} else {
```

Comment:
Suggest checking for blank value first and return error and then do rest of processing.